packages/query-core/src/__tests__/query.test.tsx

```diff
@@ -965,4 +965,44 @@ describe('query', () => {
     await sleep(60) // let it resolve
     expect(spy).toHaveBeenCalledWith('1 - 2')
   })
+
+  it('should have an error status when queryFn data is not serializable', async () => {
+    const key = queryKey()
+
+    const queryFn = vi.fn()
+
+    queryFn.mockImplementation(async () => {
+      await sleep(1000)
+
+      const data: Array<{
+        id: number
+        name: string
+        link: null | { id: number; name: string; link: unknown }
+      }> = Array.from({ length: 5 })
+        .fill(null)
+        .map((_, index) => ({
+          id: index,
+          name: `name-${index}`,
+          link: null,
+        }))
+
+      if (data[0] && data[1]) {
+        data[0].link = data[1]
+        data[1].link = data[0]
+      }
+
+      return data
+    })
+
+    await queryClient.prefetchQuery({ queryKey: key, queryFn })
+
+    const query = queryCache.find({ queryKey: key })!
+
+    expect(queryFn).toHaveBeenCalledTimes(1)
+
+    // NOTE: need second query fetch, cause for first fetch 'prevData' for structural sharing is 'undefined'
+    await query.fetch()
```

Comment:
this is weird. are you saying the query goes in to success state first, and only after a refetch, it will go into error state ?

Reply:
yes, it goes in a success state because on the first prefetch, this condition is not `true`: https://github.com/TanStack/query/blob/main/packages/query-core/src/utils.ts#L251
Because `this.state.data` will be `undefined` for the first prefetch: https://github.com/TanStack/query/blob/main/packages/query-core/src/query.ts#L215

Reply:
right, makes sense. I want to avoid having to do a safety check like you showed with `try JSON.stringify`, because it would add and additional serialization step for everyone who sends "normal" data.

What we can do is add this check to `replaceEqualDeep` behind an env check, like we do here: 

https://github.com/TanStack/query/blob/bd3296bfed9b4db1882e42fe378f2f35dc240381/packages/query-core/src/hydration.ts#L85-L89

That would also make sure that we don't need two fetches in the test. Yes, in production it would still work for the first fetch and only error on the second, but I think that's okay

Reply:
kk, will implement

Reply:
@TkDodo please have a look when you have time